File: Sources/CountriesFlagsObjC/source/Countries/UIImage+SaoTomeAndPrincipe.m
//
//  UIImage+SaoTomeAndPrincipe.m
//  CountriesFlags
//

#import "Countries/UIImage+SaoTomeAndPrincipe.h"

#import "Extensions/UIBezierPath+Triangle.h"
#import "Extensions/UIBezierPath+Star.h"

@implementation UIImage (SaoTomeAndPrincipe)

+ (UIImage * _Nullable)flagSaoTomeAndPrincipeWithSize:(CGSize)size {
    UIColor* colorGreen = [UIColor colorWithRed:0.0f
                                          green:151.0f/255.0f
                                           blue:57.0f/255.0f
                                          alpha:1.0f];
    UIColor* colorYellow = [UIColor colorWithRed:1.0f
                                           green:209.0f/255.0f
                                            blue:0.0f
                                           alpha:1.0f];
    UIColor* colorBlack = [UIColor blackColor];
    UIColor* colorRed = [UIColor colorWithRed:239.0f/255.0f
                                        green:51.0f/255.0f
                                         blue:64.0f/255.0f
                                        alpha:1.0f];
    
    UIGraphicsBeginImageContextWithOptions(size, NO, 0.0f);
    
    [colorGreen setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorYellow setFill];
    UIRectFill(CGRectMake(0.0f, size.height / 84.0f * 24.0f, size.width, size.height / 84.0f * 36.0f));
    
    [colorRed setFill];
    [[UIBezierPath triangleInSize:CGSizeMake(size.width / 168.0f * 84.0f, size.height)
                             type:FlagTriangleTypeLeftMiddle] fill];
    
    [colorBlack setFill];
    CGFloat starRadius = size.height / 84.0f * 28.74489303 / 2.0f;
    CGPoint starCenter = CGPointMake(size.width / 2.0f, (size.height + size.height / 84.0f * 1.37244652 * 2.0f) / 2.0f);
    [[UIBezierPath starAtPoint:starCenter
                        radius:starRadius
                          type:FlagStarTypeFivePointed] fill];
    starCenter = CGPointMake(size.width * 1.5f / 2.0f, (size.height + size.height / 84.0f * 1.37244652 * 2.0f) / 2.0f);
    [[UIBezierPath starAtPoint:starCenter
                        radius:starRadius
                          type:FlagStarTypeFivePointed] fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image;
}

@end

File: Sources/CountriesFlags/Countries/UIImage+SaoTomeAndPrincipe.swift
//
//  UIImage+SaoTomeAndPrincipe.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_São_Tomé_and_Príncipe "wikipedia"
     
     Generates the flag of São Tomé and Príncipe.
     For more information see [wikipedia].

     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of São Tomé and Príncipe.
     */
    internal static func flagSaoTomeAndPrincipe(with size: CGSize) -> UIImage? {
        let colorGreen = UIColor(red: 0.0,
                                 green: 151.0 / 255.0,
                                 blue: 57.0 / 255.0,
                                 alpha: 1.0)
        let colorYellow = UIColor(red: 1.0,
                                  green: 209.0 / 255.0,
                                  blue: 0.0,
                                  alpha: 1.0)
        let colorBlack = UIColor.black
        let colorRed = UIColor(red: 239.0 / 255.0,
                               green: 51.0 / 255.0,
                               blue: 64.0 / 255.0,
                               alpha: 1.0)

        UIGraphicsBeginImageContextWithOptions(size, false, 0.0)

        colorGreen.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorYellow.setFill()
        UIRectFill(CGRect(origin: CGPoint(x: 0.0, y: size.height / 84.0 * 24.0),
                          size: CGSize(width: size.width, height: size.height / 84.0 * 36.0)))

        colorRed.setFill()
        UIBezierPath.triangle(in: CGSize(width: size.width / 168.0 * 42.0, height: size.height),
                              type: .left).fill()

        colorBlack.setFill()
        var starSize = CGSize(width: size.width, height: size.height + size.height / 84.0 * 1.37244652 * 2.0)
        UIBezierPath.star(with: starSize,
                          radius: size.height / 84.0 * 28.74489303 / 2.0).fill()
        starSize = CGSize(width: size.width * 1.5, height: size.height + size.height / 84.0 * 1.37244652 * 2.0)
        UIBezierPath.star(with: starSize,
                          radius: size.height / 84.0 * 28.74489303 / 2.0).fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image
    }
}

#endif
